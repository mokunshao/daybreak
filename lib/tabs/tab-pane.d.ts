import React, { HTMLProps, ReactNode } from 'react';
interface Props extends HTMLProps<HTMLDivElement> {
    tab: ReactNode;
}
export declare const TabPane: React.FC<Props>;
export default TabPane;
//# sourceMappingURL=tab-pane.d.ts.map