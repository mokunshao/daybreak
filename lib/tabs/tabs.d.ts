import React, { HTMLProps } from 'react';
import './tabs.scss';
export declare const baseClass: (...names: (string | undefined)[]) => string;
interface Props extends Omit<HTMLProps<HTMLDivElement>, 'onChange'> {
    active: number;
    onChange: (i: number) => void;
    isVertical?: boolean;
}
export declare const Tabs: React.FC<Props>;
export default Tabs;
//# sourceMappingURL=tabs.d.ts.map