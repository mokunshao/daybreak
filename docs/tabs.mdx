---
title: Tabs
---

import {useState} from 'react';
import { Playground, Props } from 'docz';
import Tabs from '../src/tabs/tabs.tsx';
import TabPane from '../src/tabs/tab-pane.tsx';
import TabsE from './tabs.tsx';

# Tabs

```javascript
import { Tabs } from 'daybreak';
```

## Basic usage

<Playground>
  <TabsE/>
</Playground>

## Props

### Tabs


### Tab

