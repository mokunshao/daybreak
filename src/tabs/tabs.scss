@import '../common.scss';

.daybreak-tabs {
  display: flex;
  flex-direction: column;

  &-line {
    position: absolute;
    background-color: black;
    transition: all 0.4s;
  }

  &-tab {
    padding: 0.3em;
    cursor: pointer;

    &:hover {
      background-color: $grey2;
    }
  }

  &-list {
    background-color: $grey;
    position: relative;
    display: inline-flex;
  }

  &-panes {
    flex: 1;
  }

  &-vertical {
    flex-direction: row;
  }

  &-vertical &-list {
    flex-direction: column;
  }
}
