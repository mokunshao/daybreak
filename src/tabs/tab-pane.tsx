import React, { HTMLProps, ReactNode } from 'react';
import { classes } from '../utils/classes';
import { baseClass } from './tabs';

interface Props extends HTMLProps<HTMLDivElement> {
  tab: ReactNode;
}

export const TabPane: React.FC<Props> = (props) => {
  const {
    children, className, ...rest
  } = props;
  return (
    <div className={classes(baseClass('pane'), className)} {...rest}>
      {children}
    </div>
  );
};

export default TabPane;
