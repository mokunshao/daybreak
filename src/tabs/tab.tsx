import React, { HTMLProps } from 'react';
import { classes } from '../utils/classes';
import { baseClass } from './tabs';

interface Props extends HTMLProps<HTMLDivElement> {
  title: string;
}

export const Tab: React.FC<Props> = (props) => {
  const {
    children, className, ...rest
  } = props;
  return (
    <div className={classes(baseClass('pane'), className)} {...rest}>
      {children}
    </div>
  );
};

export default Tab;
