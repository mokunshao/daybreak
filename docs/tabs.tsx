import React, { useState } from 'react';
import { Tabs } from '../src/tabs/tabs';
import { TabPane } from '../src/tabs/tab-pane';

export default () => {
  const [active, setActive] = useState(0);
  const [active1, setActive1] = useState(0);
  return (
    <div>
      <Tabs active={active} onChange={(index) => setActive(index)}>
        <TabPane tab='1121999999'>Tab 1</TabPane>
        <TabPane tab="okokkk">Tab 2</TabPane>
      </Tabs>
      <br />
      <Tabs isVertical active={active1} onChange={(index) => setActive1(index)}>
        <TabPane tab='1121'>Tab 1</TabPane>
        <TabPane tab="okok">Tab 2</TabPane>
      </Tabs>
    </div>
  );
};