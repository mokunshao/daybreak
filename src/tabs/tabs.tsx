import React, {
  HTMLProps, ReactElement, useMemo, useRef, useCallback, useEffect, RefObject,
} from 'react';
import { classes } from '../utils/classes';
import { joinedClass } from '../utils/joinedClass';
import './tabs.scss';
import { TabPane } from './tab-pane';

export const baseClass = joinedClass('tabs');

interface Props extends Omit<HTMLProps<HTMLDivElement>, 'onChange'> {
  active: number;
  onChange: (i: number) => void;
  isVertical?: boolean;
}

export const Tabs: React.FC<Props> = React.memo((props) => {
  const {
    children, className, active = 0,
    onChange, isVertical = false, ...rest
  } = props;

  const TheChildren = useRef(children);

  const tabsList = useMemo(() => {
    const arr: Array<ReactElement> = [];
    React.Children.forEach(TheChildren.current, (c) => {
      if (React.isValidElement(c) && c.type === TabPane) {
        arr.push(c);
      }
    });
    return arr;
  }, [TheChildren]);

  const tabsLabel = useMemo(
    () => tabsList.map((c, i) => (
      <div ref={React.createRef()} className={baseClass('tab')} key={i} onClick={() => onChange(i)}>
        {c.props.tab}
      </div>
    )),
    [onChange, tabsList],
  );

  const lineRef = useRef(null);

  useEffect(() => {
    const { current } = lineRef as RefObject<HTMLElement>;
    if (current) {
      const { ref } = tabsLabel[active] as any;
      const rect = ref.current.getBoundingClientRect();
      if (!isVertical) {
        current.style.top = `${rect.height}px`;
        current.style.width = `${rect.width}px`;
        current.style.left = `${ref.current.offsetLeft}px`;
      }
    }
  }, [active, isVertical, tabsLabel]);

  const ActiveTab = useMemo(
    () => tabsList[active],
    [active, tabsList],
  );

  return (
    <div className={classes(baseClass(), isVertical && baseClass('vertical'), className)} {...rest}>
      <div className={baseClass('list')} style={{ position: 'relative' }}>
        {tabsLabel}
      </div>
      <div ref={lineRef} className={baseClass('line')} />
      <div className={baseClass('panes')}>
        {ActiveTab}
      </div>
    </div>
  );
});

export default Tabs;
