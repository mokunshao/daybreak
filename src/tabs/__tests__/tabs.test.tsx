import React, { useState } from 'react';
import { render } from '@testing-library/react';
import { Tabs } from '../tabs';
import { TabPane } from '../tab-pane';

function Example() {
  const [active, setActive] = useState(1);
  return (
    <div>
      <Tabs isVertical active={active} onChange={(index) => setActive(index)}>
        <TabPane tab="1121">Tab 1 Content</TabPane>
        <TabPane tab="okok">Tab 2 Content</TabPane>
      </Tabs>
    </div>
  );
}


describe('Tabs', () => {
  it('exits', () => {
    expect(Tabs).toBeTruthy();
    expect(TabPane).toBeTruthy();
    expect(<Example />).toBeTruthy();
  });

  it('match snapshot', () => {
    const { asFragment } = render(<Example />);
    expect(asFragment()).toMatchSnapshot();
  });
});
