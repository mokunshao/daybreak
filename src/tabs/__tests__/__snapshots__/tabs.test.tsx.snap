// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs match snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="daybreak-tabs daybreak-tabs-vertical"
    >
      <div
        class="daybreak-tabs-list"
      >
        <div
          class="daybreak-tabs-tab"
        >
          1121
        </div>
        <div
          class="daybreak-tabs-tab"
          style="background-color: rgb(226, 226, 226);"
        >
          okok
        </div>
        <div
          class="daybreak-tabs-line"
          style="left: -2px; height: 0px; width: 2px; top: 0px;"
        />
      </div>
      <div
        class="daybreak-tabs-panes"
      >
        <div
          class="daybreak-tabs-pane"
          tab="okok"
        >
          Tab 2 Content
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
